compiler/optimizer/LoopVersioner.cpp

```diff
@@ -4980,15 +5005,40 @@ void TR_LoopVersioner::versionNaturalLoop(TR_RegionStructure *whileLoop, List<TR
          bool isExceptionEdge = false;
          if (exitEdge->getFrom()->hasExceptionSuccessor(node))
             isExceptionEdge = true;
-         properRegion->addExitEdge(whileNode, node->getNumber(), isExceptionEdge);
+
+         // Check whether this exit is from the while loop to the new region
+         // If it is, it should not be added as an exit from the region itself,
+         // but rather as an edge from the while loop to the start of the new
+         // region
+         if (node->getNumber() == regionEntryNode->getNumber())
+            {
+            TR::CFGEdge::createEdge(whileNode, regionEntryNode, trMemory());
+            }
+         else
+            {
+            properRegion->addExitEdge(whileNode, node->getNumber(), isExceptionEdge);
+            }
          seenExitNodes.set(node->getNumber());
          }
       }
 
+   // If the original while loop exited to the loop invariant block, ensure the
+   // structure of the cloned version of the loop exits to the new proper region
+   ei.set(&clonedWhileLoop->getExitEdges());
+   for (exitEdge = ei.getCurrent(); exitEdge; exitEdge = ei.getNext())
+      {
+      if (exitEdge->getTo()->getNumber() == invariantBlockStructure->getNumber())
+         {
+if (trace()) traceMsg(comp(), "HZ - calling replaceExitPart for structure %p from %d to %d\n", clonedWhileLoop, invariantBlockStructure->getNumber(), properRegion->getNumber());
```

Comment:
This is highly unusual formatting. Did you mean to remove the tracing? If not, please reformat

Reply:
And please don't initial your messages.  If they're for your own private testing then they shouldn't be part of this PR.

Reply:
Yikes!  I hadn't intended to push it with that trace message there.  Sorry about that.

Reply:
Devin @jdmpapin, @0xdaryl Daryl, I removed that stray trace message and squashed that entire revision into the first commit https://github.com/eclipse/omr/pull/3745/commits/0c23e91ebcc9ae0c3f23f8976cf9120db987c17a